txnkv/transaction/prewrite.go

```diff
@@ -226,286 +227,368 @@ func (action actionPrewrite) handleSingleBatch(
+		// TiKV returns lock.TTL = 0, and we still need to resolve the lock.
+		if lock.TxnID > handler.committer.startTS && !handler.committer.isPessimistic {
+			return nil, tikverr.NewErrWriteConflictWithArgs(
+				handler.committer.startTS,
+				lock.TxnID,
+				0,
+				lock.Key,
+				kvrpcpb.WriteConflict_Optimistic,
+			)
 		}
-		resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+		locks = append(locks, lock)
+	}
+	return locks, nil
+}
+
+// resolveLocks resolves locks.
+func (handler *prewrite1BatchReqHandler) resolveLocks(locks []*txnlock.Lock) error {
+	if handler.resolvingRecordToken == nil {
+		token := handler.committer.store.GetLockResolver().RecordResolvingLocks(locks, handler.committer.startTS)
+		handler.resolvingRecordToken = &token
+	} else {
+		handler.committer.store.GetLockResolver().UpdateResolvingLocks(locks, handler.committer.startTS, *handler.resolvingRecordToken)
+	}
+
+	resolveLockOpts := txnlock.ResolveLocksOptions{
+		CallerStartTS:            handler.committer.startTS,
+		Locks:                    locks,
+		Detail:                   &handler.committer.getDetail().ResolveLock,
+		PessimisticRegionResolve: true,
+	}
+	resolveLockRes, err := handler.committer.store.GetLockResolver().ResolveLocksWithOpts(handler.bo, resolveLockOpts)
+	if err != nil {
+		return err
+	}
+	msBeforeExpired := resolveLockRes.TTL
+	if msBeforeExpired > 0 {
+		err = handler.bo.BackoffWithCfgAndMaxSleep(
+			retry.BoTxnLock,
+			int(msBeforeExpired),
+			errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)),
+		)
 		if err != nil {
 			return err
 		}
-		msBeforeExpired := resolveLockRes.TTL
-		if msBeforeExpired > 0 {
-			err = bo.BackoffWithCfgAndMaxSleep(
-				retry.BoTxnLock,
-				int(msBeforeExpired),
-				errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)),
+	}
+	return nil
+}
+
+// handleSingleBatchSucceed handles the response when the prewrite request is successful.
+func (handler *prewrite1BatchReqHandler) handleSingleBatchSucceed(prewriteResp *kvrpcpb.PrewriteResponse) error {
+	// Clear the RPC Error since the request is evaluated successfully.
+	handler.sender.SetRPCError(nil)
+
+	// Update CommitDetails
+	reqDuration := time.Since(handler.begin)
+	handler.committer.getDetail().MergePrewriteReqDetails(
+		reqDuration,
+		handler.batch.region.GetID(),
+		handler.sender.GetStoreAddr(),
+		prewriteResp.ExecDetailsV2,
+	)
+
+	if handler.batch.isPrimary {
+		// After writing the primary key, if the size of the transaction is larger than 32M,
+		// start the ttlManager. The ttlManager will be closed in tikvTxn.Commit().
+		// In this case 1PC is not expected to be used, but still check it for safety.
+		if int64(handler.committer.txnSize) > config.GetGlobalConfig().TiKVClient.TTLRefreshedTxnSize &&
+			prewriteResp.OnePcCommitTs == 0 {
+			handler.committer.run(handler.committer, nil, false)
```

Comment:
Though `ttlManager` is embed in `twoPhaseCommitter`, call `committer.run` is little weird. Better to change it to:

```suggestion
			handler.ttlManager.committer.run(handler.committer, nil, false)
```